### PR TITLE
[i1] Remove command line option to enable packed storage

### PR DESCRIPTION
**Pull request by lialan**
* only use `#iree_encoding.packed_storage` to designate if an `i1` tensor is of packed memory layout.
* remove `iree-experimental-packed-i1-storage` command line option.
* teach type converters to allow casting into packed tensor types

### CODE REVIEW

**Comment by lialan** [comment]
This is a successor, improving PR to https://github.com/iree-org/iree/pull/19354

**Comment by hanhanW** [comment]
The diff is off. There are some changes from https://github.com/iree-org/iree/pull/19354. E.g., the below is not part of this PR, right?

image

Can you rebase and fix it?

**Comment by lialan** [comment]
Note that changes that are not in this PR: #19618

File: compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp [hunk 1]
```diff
@@ -80,6 +83,14 @@ MaterializeEncodingConversionTarget::MaterializeEncodingConversionTarget(
 }
 
 RankedTensorType dropEncoding(RankedTensorType type) {
+  assert(!IREE::Encoding::hasPackedStorageAttr(type) &&
+         "not expected `packed_storage` attribute.");
+  return RankedTensorType::get(type.getShape(), type.getElementType());
+}
+
+RankedTensorType dropPackedStorageEncodingIfAny(RankedTensorType type) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp:103]: Why do we have two dropping encoding methods? Can we just have one? If we just look at the method name, `dropEncoding` should drop all the encodings.

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.h [hunk 2]
```diff
@@ -30,6 +30,8 @@ bool needToPackSubByteElements(RankedTensorType shapedType);
 /// cases.
 Type legalizeStorageElementType(Type elementType);
 
+Type legalizeTensorStorageElementType(Type tensorType);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Utils/ElementPackingUtils.h:33]: IIRC, this is a rename of `legalizeStorageElementType`, and the implementation of `legalizeStorageElementType` is deleted. Can we just use `legalizeStorageElementType` without renaming? Also, the `legalizeStorageElementTypeImpl` seems no longer needed, can we remove the method, i.e., replace `legalizeStorageElementTypeImpl` with `legalizeStorageElementType`?

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp [hunk 3]
```diff
@@ -79,10 +59,10 @@ static Type legalizeStorageElementTypeImpl(Type elementType,
                           intType.getSignedness());
 }
 
-Type legalizeStorageElementType(Type elementType) {
-  // Consider packed storage for i1 tensors if cl opt is set.
-  return legalizeStorageElementTypeImpl(elementType,
-                                        /*isPackedStorage=*/clEnableI1Support);
+Type legalizeTensorStorageElementType(Type type) {
+  auto tensorType = llvm::cast<TensorType>(type);
+  return legalizeStorageElementTypeImpl(
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp:64]: I think we can move the logic of `legalizeStorageElementTypeImpl` to here, because this is the only user, right?

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp [hunk 4]
```diff
@@ -97,15 +77,16 @@ Value calculateStorageElementCountInBytes(Location loc,
...
   unsigned elementBits = IREE::Util::getTypeBitWidth(alignedElementType);
 
+  bool isPackedStorage = IREE::Encoding::hasPackedStorageAttr(shapedType);
+  bool isI1WithPackedStorage = elementBits == 1 && isPackedStorage;
+
   // Calculate all static dims first, if any.
   int64_t staticCount = 1;
-  if (!needToPackSubByteElementBitWidthImpl(elementBits, isPackedStorage)) {
+  if (!isI1WithPackedStorage &&
+      !needToPackSubByteElementBitWidth(elementBits)) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp:88]: Perhaps we should pass the type to `needToPackSubByteElementBitWidth`, then we don't need `isI1WithPackedStorage` everywhere.

The other use is in EncodeTensor.cpp. Looking at the method, I think it is easy to update the logic, i.e., just pass the type but not `elementBits`.

https://github.com/iree-org/iree/blob/a583b25df17873b9ae214041ba107e47d00f22e1/compiler/src/iree/compiler/Dialect/Stream/Transforms/EncodeTensors.cpp#L155-L171

File: tests/e2e/subbyte_types/subbyte_types_attr.mlir [hunk 5]
```diff
@@ -0,0 +1,56 @@
+#packed = #iree_encoding.packed_storage
+func.func @i1_type_slice() {
+  %input = util.unfoldable_constant dense<[0, 255, 0]> : tensor<3xi8>
+  %flat_input_all = flow.tensor.bitcast %input : tensor<3xi8> -> tensor<24xi1, #packed>
```
Comment (hanhanW) [inline tests/e2e/subbyte_types/subbyte_types_attr.mlir:4]: I'm not familiar with how `flow.tensor.bitcast` works. Is the encoding dropped in the `TensorExportBufferViewOpPattern` pattern? How does it become a tensor_export_buffer_view op? I think we need some input from Ben about the change and the test.

**Review by hanhanW — CHANGES_REQUESTED** [review]
First round of the review comments, thanks for pushing on this!

File: tests/e2e/subbyte_types/subbyte_types_attr.mlir (hunk 5 above)
Comment (lialan) [reply tests/e2e/subbyte_types/subbyte_types_attr.mlir:4]: The encoding dropped while it is calculating streamOp's allocated sizes.

`flow.tensor.bitcast` is just an omni-potent caster here to cast a tensor to another tensor with `#packed_storage` attribute.

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp (hunk 3 above)
Comment (lialan) [reply compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp:64]: yap, you are right. it was just some residual for my previous refactoring. Thanks for pointing this out!

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.h (hunk 2 above)
Comment (lialan) [reply compiler/src/iree/compiler/Utils/ElementPackingUtils.h:33]: done.

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp (hunk 4 above)
Comment (lialan) [reply compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp:88]: There is one function where the element type and the shape type are handled separately, specifically here: https://github.com/iree-org/iree/pull/19528/files#diff-9ccfc5b598b4e78bd96d52bb2d98f6a07d2d4c34803f80fe3a15248c8700b7a0L102-R75

Still I have made the function to take type instead of bitwidth. But that does not completely eliminate `isI1WithPackedStorage` in this function.

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.td [hunk 6]
```diff
@@ -17,49 +17,53 @@ include "mlir/IR/EnumAttr.td"
...
+                               ]>;
+
+def EncodingOpTypeAttr : IREEEncoding_EnumAttr<EncodingOpType, "optype">;
+
+def PackedStorageAttr
+    : IREEEncoding_Attr<"PackedStorage",
+                        [DeclareAttrInterfaceMethods<
+                            IREEEncoding_EncodingLayoutAttrInterface, [
+                              "calculateStorageSizeInBytes",
+                            ]>]> {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.td]: I don't expect changes other than this part in this file. Can you revert other changes?

File: compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp [hunk 7]
```diff
@@ -478,7 +479,8 @@ struct TensorExportBufferViewOpPattern
     }
 
     auto loc = exportOp.getLoc();
-    auto tensorType = llvm::cast<RankedTensorType>(adaptor.getSourceEncoding());
+    auto tensorType =
+        dropEncoding(llvm::cast<RankedTensorType>(adaptor.getSourceEncoding()));
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp]: nit: need a comment about why we drop the encoding. If it were me, I'd write:

```cpp
auto tensorType = llvm::cast<RankedTensorType>(adaptor.getSourceEncoding());
// why we drop the encoding
tensorType = dropEncoding(tensorType)
```

File: compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp [hunk 8]
```diff
@@ -27,6 +28,10 @@
 #include "mlir/Support/LogicalResult.h"
 #include "mlir/Transforms/RegionUtils.h"
 
+namespace mlir::iree_compiler {
+using IREE::Encoding::getEncodingAttr;
+}
+
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp:34]: We do not need this. They are declared in the `EncodingTypes.h` and you already include it.

File: compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp [hunk 9]
```diff
@@ -1512,7 +1517,7 @@ LogicalResult TensorCloneOp::verify() {
   // information.
   auto sourceEncoding = llvm::cast<RankedTensorType>(op.getSourceEncoding());
   auto resultEncoding = llvm::cast<RankedTensorType>(op.getResultEncoding());
-  if (sourceEncoding.getEncoding() != resultEncoding.getEncoding()) {
+  if (getEncodingAttr(sourceEncoding) != getEncodingAttr(resultEncoding)) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp:1911]: Is this required by `flow.tensor.bitcast` lowering? I.e., the op is lowered to flow.tensor.clone and you need to bypass the check?

I'm not convinced that the change is correct. Because the `getEncodingAttr` checks if the tensor type has `IREE::Encoding::EncodingAttr` attribute. There could be other encodings and it is a bug if we introduce new encodings. E.g., `tensor<3x4xi32, #whatever_other_encoding_with_padding_semantic>` can not be cloned to `tensor<3x4xi32>`. I think we need to have a stronger restriction. Perhaps just relax the check for packed_storage encoding?

https://github.com/iree-org/iree/blob/27e7a9058eded6a20dc098368352329263f77a23/compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp#L278-L280

File: compiler/src/iree/compiler/Dialect/Stream/Transforms/ConvertToStream.cpp [hunk 10]
```diff
@@ -247,6 +249,12 @@ struct ConvertToStreamPass final
       if (llvm::isa<IREE::Flow::ChannelType>(type)) {
         return IREE::Stream::ChannelType::get(context);
       }
+      if (auto rankedType = llvm::dyn_cast_or_null<RankedTensorType>(type)) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Stream/Transforms/ConvertToStream.cpp:252]: Do we really need to check if the incoming type is null type or not? If not, we can just use `dyn_cast`.

Also, please add a comment for the below check.

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp [hunk 11]
```diff
@@ -15,25 +15,19 @@
...
-  if (isPackedStorage && bitWidth == 1) {
+bool needToPackSubByteElements(Type type) {
+  unsigned bitWidth = isa<TensorType>(type)
+                          ? IREE::Util::getTypeBitWidth(
+                                dyn_cast<TensorType>(type).getElementType())
+                          : IREE::Util::getTypeBitWidth(type);
+
+  auto rankedTensorType = llvm::dyn_cast_or_null<RankedTensorType>(type);
+  bool isPackedStorage = rankedTensorType &&
+                         IREE::Encoding::hasPackedStorageAttr(rankedTensorType);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp]: Use dyn_cast because it is never a null type. Also, I think we can simplify it a bit, like:

```cpp
auto rankedTensorType = dyn_cast...
if (!rankedTensorType) {
  return false;
}
unsigned bitWidth = IREE::Util::getTypeBitWidth(rankedTensorType.getElementType());
if (bitWidth == 1 && IREE::Encoding::hasPackedStorageAttr...
```

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp [hunk 12]
```diff
@@ -152,6 +152,45 @@ static int32_t getRoundedElementByteWidth(Type type) {
...
+    Location loc, OpBuilder &builder, RankedTensorType type,
+    ValueRange dynamicDims) const {
+  unsigned elementBits = type.getElementTypeBitWidth();
+  assert(elementBits == 1 && "packed storage only supported for i1 for now");
+
+  // Calculate all static dims first, if any.
+  bool isPackedStorage = IREE::Encoding::hasPackedStorageAttr(type);
+  int64_t staticCount = 1;
+  if (!isPackedStorage) {
+    staticCount *= elementBits * 8;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp]: Let's avoid magic number: We can move `constexpr int kNumBitsInByte = 8` to global and use it in the file.

https://github.com/iree-org/iree/blob/27e7a9058eded6a20dc098368352329263f77a23/compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp#L217

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp [hunk 13]
```diff
@@ -152,6 +152,45 @@ static int32_t getRoundedElementByteWidth(Type type) {
...
+    unsigned byteElements = 8 / elementBits;
+    // TODO(antiagainst): We may want to emit runtime check to make sure this is
+    // divisible.
+    auto divisor = builder.create<arith::ConstantIndexOp>(loc, byteElements);
+    if (!isPackedStorage && dynamicDims.empty() &&
+        (staticCount * elementBits) % 8 != 0) {
+      return nullptr;
+    }
+    value = builder.createOrFold<arith::CeilDivUIOp>(loc, value, divisor);
+  }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp]: I think we can simplify this part a lot because we already assert if elementBits is 1 or not in the beginning. My mental model is keeping code simple, and extend it when needed. We definitely need to balance the mental model, but this part is much more complicated than the current need IMO.

btw, I recently learned that we prefer using CeilDivSIOp unless there is a specific reason. See https://github.com/iree-org/iree/pull/19686#discussion_r1915031028 for more details.

**Review by hanhanW — CHANGES_REQUESTED** [review]
Thanks for the update! Please also add a test to https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/test/type_propagation_packing.mlir

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp (hunk 12 above)
Comment (lialan) [reply compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp]: Great, I've updated all related parts in this file.

File: compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp (hunk 9 above)
Comment (lialan) [reply compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp:1520]: I think you are right. I added extra test for the packed attribute.

Comment (lialan) [reply compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp:1911]: This is a little tricky. And yes this is required by `flow.tensor.bitcast`, and we would sometimes cast from a tensor without attribute to another tensor with packed attribute. In such case, we shouldn't check if both the source and result matches packed attribute.

I have slightly updated it to exclude packed attribute comparison. Suggestions are welcome.

File: compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp [hunk 14]
```diff
@@ -29,10 +29,13 @@ MaterializeEncodingTypeConverter::MaterializeEncodingTypeConverter(
   addConversion([](FloatType floatType) { return floatType; });
   addConversion([](MemRefType memrefType) { return memrefType; });
   addConversion([=](RankedTensorType type) -> RankedTensorType {
+    MaterializeEncodingInfo encodingInfo = getEncodingInfo(type);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp:33]: Sorry, I missed this in the previous review. Please move it back to where it was. It is not used until l.39 and the below check does not need this information. It also matches the comments better.

(The below comment is a little outdated, I'll fix them in a separate PR.)

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp [hunk 15]
```diff
@@ -188,8 +221,11 @@ Value EncodingAttr::calculateStorageSizeInBytes(Location loc,
     pad(k, roundDimsTo[2]);
   }
 
-  constexpr int64_t kNumBitsInByte = 8;
   unsigned elementBits = getTypeBitWidth(type.getElementType());
+  // Deal with unpacked storage of i1.
+  if (elementBits == 1 && !IREE::Encoding::hasPackedStorageAttr(type)) {
+    elementBits = kNumBitsInByte;
+  }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp]: Just a note. The change is fine. In the future, the logic will be moved to `Codegen/ExternalInterfaces/Utils.cpp`. Now we are in a transition state. It will happen after we finish the encoding specialization pass.

https://github.com/iree-org/iree/blob/3c95042d82a310bedc67db2e065680dd85c58d96/compiler/src/iree/compiler/Codegen/ExternalInterfaces/Utils.cpp#L81-L90

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingTypes.h [hunk 16]
```diff
@@ -39,7 +39,7 @@ namespace mlir::iree_compiler::IREE::Encoding {
 EncodingAttr getEncodingAttr(RankedTensorType type);
 
 /// Returns true if the type contains packed_storage attribute.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingTypes.h:45]: nit:

```suggestion
/// Returns true if the type is a RankedTensorType type and it contains packed_storage attribute.
```

File: compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp [hunk 17]
```diff
@@ -6,6 +6,7 @@
 
 #include "iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.h"
 
+#include "iree/compiler/Codegen/Common/EncodingUtils.h"
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp:9]: We do not want this dependency. I think you want to reuse `dropEncoding` method. Let's move it to `Encoding/EncodingTypes.h` and `EncodingAttr.cpp`. 

(Ideally we should move the implementation to `EncodingTypes.cpp`. I haven't started the work because we have few people touching the dialect at this moment. I'll send a cleanup after you finish the i1 work.)

File: compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp [hunk 18]
```diff
@@ -478,7 +479,10 @@ struct TensorExportBufferViewOpPattern
     }
 
     auto loc = exportOp.getLoc();
-    auto tensorType = llvm::cast<RankedTensorType>(adaptor.getSourceEncoding());
+    // Drop the encoding of packed_storage here, as it is no longer needed
+    // afterwards.
+    auto tensorType =
+        dropEncoding(llvm::cast<RankedTensorType>(adaptor.getSourceEncoding()));
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp]: I'm confused. Won't this crash in debug build? We have below assertion in the implementation:

```
assert(!IREE::Encoding::hasPackedStorageAttr(type) &&
         "not expected `packed_storage` attribute.");
```

File: compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp [hunk 19]
```diff
@@ -1512,7 +1513,9 @@ LogicalResult TensorCloneOp::verify() {
   // information.
   auto sourceEncoding = llvm::cast<RankedTensorType>(op.getSourceEncoding());
   auto resultEncoding = llvm::cast<RankedTensorType>(op.getResultEncoding());
-  if (sourceEncoding.getEncoding() != resultEncoding.getEncoding()) {
+  if (sourceEncoding.getEncoding() != resultEncoding.getEncoding() &&
+      !IREE::Encoding::hasPackedStorageAttr(sourceEncoding) &&
+      !IREE::Encoding::hasPackedStorageAttr(resultEncoding)) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp]: It looks better to me. I think we need some input from @benvanik 

The previous comment is not tracked in the review mode. Here was the previous comment: https://github.com/iree-org/iree/pull/19528#discussion_r1915955126 

TLDR is that we use flow.tensor.bitcast to prepare packed data for e2e testing. The op becomes TensorClone op during the lowering. So we want to relax the verification here.

```
%flat_input_all = flow.tensor.bitcast %input : tensor<3xi8> -> tensor<24xi1, #packed>
```

**Review by hanhanW — CHANGES_REQUESTED** [review]
Reminder: the test is not added yet. https://github.com/iree-org/iree/pull/19528#pullrequestreview-2551569009

File: compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp (hunk 18 above)
Comment (lialan) [reply compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp]: You are right, this is some code left in previous iteration. removed it and its deps.

File: compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp (hunk 17 above)
Comment (lialan) [reply compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp:9]: removed all those deps.

**Comment by lialan** [comment]
@hanhanW I've added a test to test type propagation.

File: compiler/src/iree/compiler/Codegen/Common/test/type_propagation_packing.mlir [hunk 20]
```diff
@@ -34,3 +34,43 @@ func.func @generic_op_i4() {
...
+//   CHECK-DAG:   %[[INTENSOR:.+]] = flow.dispatch.tensor.load %[[IN]]{{.+}} -> tensor<?xi1, #iree_encoding.packed_storage>
+//   CHECK-DAG:   %[[INIT:.+]] = tensor.empty(%{{.+}}) : tensor<?xi1, #iree_encoding.packed_storage>
+//       CHECK:   %[[GENERIC:.+]] = linalg.generic
+//  CHECK-SAME:       ins(%[[INTENSOR]] : tensor<?xi1, #iree_encoding.packed_storage>)
+//  CHECK-SAME:       outs(%[[INIT]] : tensor<?xi1, #iree_encoding.packed_storage>)
+//  CHECK-NEXT:     ^bb0(%[[ARG0:[a-zA-Z0-9]+]]: i1, %[[ARG1:[a-zA-Z0-9]+]]: i1)
+//   CHECK-DAG:       %[[ADD:.+]] = arith.addi %[[ARG0]], %[[ARG0]] : i1
+//       CHECK:       linalg.yield %[[ADD]]
+//       CHECK:   -> tensor<?xi1, #iree_encoding.packed_storage>
+//       CHECK:   flow.dispatch.tensor.store %[[GENERIC]], %[[OUT]]
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/test/type_propagation_packing.mlir]: nit: missing new line at the end.

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp [hunk 21]
```diff
@@ -253,8 +289,15 @@ EncodingAttr getEncodingAttr(RankedTensorType type) {
...
 
-bool hasPackedStorageAttr(RankedTensorType type) {
-  return dyn_cast_or_null<PackedStorageAttr>(type.getEncoding()) != nullptr;
+bool hasPackedStorageAttr(Type type) {
+  if (auto tensorType = dyn_cast<RankedTensorType>(type)) {
+    auto encoding = tensorType.getEncoding();
+    if (!encoding) {
+      return false;
+    }
+    return dyn_cast_or_null<PackedStorageAttr>(encoding) != nullptr;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp]: I think this can be `return dyn_cast_or_null<xxx>(tensorType.getEncoding)`?

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingTypes.h (hunk 16 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingTypes.h:45]: It is marked resolved. Is it addressed?

**Review by hanhanW — COMMENTED** [review]
Just a few final nits. LGTM except the `TensorCloneOp::verify` part. Please coordinate with @benvanik about it.

File: compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp (hunk 19 above)
Comment (lialan) [reply compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp]: @benvanik Do you have opinions on this?

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp [hunk 22]
```diff
@@ -96,15 +76,16 @@ Value calculateStorageElementCountInBytes(Location loc,
...
 
-  bool isPackedStorage =
-      IREE::Encoding::hasPackedStorageAttr(shapedType) || clEnableI1Support;
-  Type alignedElementType = legalizeStorageElementTypeImpl(
-      shapedType.getElementType(), isPackedStorage);
+  Type alignedElementType = legalizeTensorStorageElementType(shapedType);
   unsigned elementBits = IREE::Util::getTypeBitWidth(alignedElementType);
 
+  bool isPackedStorage = IREE::Encoding::hasPackedStorageAttr(shapedType);
+  bool isI1WithPackedStorage = elementBits == 1 && isPackedStorage;
```
Comment (benvanik) [inline compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp:83]: There's a lot of i1 special casing in this file still - I think it could be simplified to be `isPackedStorage = IREE::Encoding::hasPackedStorageAttr(shapedType) || needToPackSubByteElementBitWidth(elementBits)` as I can't think of a case where we'd not have those both be the case. And really, we should now be putting packed storage attrs on everything we want packed and remove needToPackSubByteElementBitWidth entirely.

File: compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp [hunk 23]
```diff
@@ -27,6 +28,10 @@
 #include "mlir/Support/LogicalResult.h"
 #include "mlir/Transforms/RegionUtils.h"
 
+namespace mlir::iree_compiler {
+using IREE::Encoding::getEncodingAttr;
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp:32]: directly reference this when required - it hurts readability to have aliases like this (as a user will just see getEncodingAttr and not know if it's on their type, in the dialect they are in, etc) - I'm not sure we want this function anywhere, though, as commented below (the flow/stream dialects should support all encodings, not just those in the encoding dialect)

File: compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp (hunk 9 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/Stream/IR/StreamOps.cpp:1911]: I agree with hanhan - I'm not sure why this changed - if it is indeed tricky it's too tricky to be as opaque as this. The op behavior and the comment both state the existing code was the check that was desired, and not that we should allow non-encoding-dialect encodings to be ignored.

File: compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp (hunk 17 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp:9]: as a general rule there are to be no codegen deps in non-codegen directories - dropPackedStorageEncodingIfAny should be moved to EncodingTypes.h

File: tests/e2e/linalg_ext_ops/BUILD.bazel [hunk 24]
```diff
@@ -42,24 +42,6 @@ iree_check_single_backend_test_suite(
     target_backend = "llvm-cpu",
 )
 
-iree_check_single_backend_test_suite(
```
Comment (benvanik) [inline tests/e2e/linalg_ext_ops/BUILD.bazel:45]: you need to un-exclude the test from the above ALL_SRCS

**Review by benvanik — CHANGES_REQUESTED** [review]

**Comment by benvanik** [comment]
Closing as stale for now (and renaming the branch to follow repository naming conventions). You can reopen/rebase if you still want to land it in the future.
